arviz/kde_utils.py

```
@@ -0,0 +1,895 @@
+    """
+    Computes bandwidth for a given data `x` and `bw`.
+    Also checks `bw` is correctly specified.
+
+    Parameters
+    ----------
+    x : 1-D numpy array
+        1 dimensional array of sample data from the
+        variable for which a density estimate is desired.
+    bw: int, float or str
+        If numeric, indicates the bandwidth and must be positive.
+        If str, indicates the method to estimate the bandwidth.
+
+    Returns
+    -------
+    bw: float
+        Bandwidth
+    """
+    if isinstance(bw, bool):
+        raise ValueError((
+            f"`bw` must not be of type `bool`.\n"
+            f"Expected a positive numeric or one of the following strings:\n"
+            f"{list(BW_METHODS_LINEAR.keys())}."))
+    if isinstance(bw, (int, float)):
+        if bw < 0:
+            raise ValueError(f"Numeric `bw` must be positive.\nInput: {bw:.4f}.")
+    elif isinstance(bw, str):
+        bw_fun = select_bw_method(bw)
+        bw_lower = bw.lower()
+        if bw_lower == "isj":
+            bw = bw_fun(x, grid_counts, x_range)
+        elif bw_lower in ["scott", "silverman"]:
+            bw = bw_fun(x, x_std)
+        elif bw_lower == "experimental":
+            bw = bw_fun(x, grid_counts, x_std, x_range)
+    else:
+        raise ValueError((
+            f"Unrecognized `bw` argument.\n"
+            f"Expected a positive numeric or one of the following strings:\n"
+            f"{list(BW_METHODS_LINEAR.keys())}."))
+    return bw
+
+
+# Misc utils ------------------------------------------------------------------------
+def normalize_angle(x, pi_centered=False):
+    """
+    Takes angles in radians and normalize them to [-pi, pi) or [0, 2 * pi)
+    depending on `pi_centered`.
+    """
+    if pi_centered:
+        return x % (2 * np.pi)
+    else:
+        return (x + np.pi) % (2 * np.pi) - np.pi
+    
+def vonmises_pdf(x, mu, kappa):
+    assert kappa > 0, "Argument 'kappa' must be positive."
```

I would use an if + raising value error here. I don't think an invalid kappa should raise an assertion error: https://docs.python.org/3/library/exceptions.html#ValueError